### DISCUSSION TITLE
null forgiving for Nullable primitive type

### DISCUSSION THREAD

**Discussion by AliRezaBeigy** [dotnet/csharplang#5203] [Q&A] [answered]
As [null-forgiving](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving) documentation, you can use ! to forgiving null value and give a not nullable type.

For example
```csharp
class A {
    public string? B;
    public int? C;
}
```
there is the A class with nullable B and C fields
when i try to use with following code, an issue showns
```csharp
    var a = new A();
    var b = a.B!; // string
    var bNullable = a.B; // Nullable<string>
    var c= a.C!; // Nullable<int>
    var cNullable = a.C; // Nullable<int>
```
as you can see, the C field always is Nullable, and null forgiving has not worked

**Comment by theunrepentantgeek** [comment]
The null-forgiving operator is _by design_ an operator that only exists at compile time. 

Transforming `int?` to `int` would require the operator to have an effect at runtime - this would be a big change.

**Comment by kyoyama-kazusa** [comment, marked as answer]
Hello.

`Nullable<int>` (known as `int?`) and `int` aren't a same type, so here `a.C` you can't use null-forgiving operator `!` to declare the not-null state on the property `C` because you may change the result value from an `int?` to an `int`. The only way to get the inner value is `int value = a.C!.Value`, or also you can use pattern matching: `if (a.C is { } value) ...`.

However, `string?` and `string` is a same type with different nullability, so `a.B!` means the value of the property `B` is not-null. Therefore the operator can be used on this case.

**Reply by jnm2** [reply]
If you do, `_ = a.C is { } value;` specifically, you can't use `value` without getting a "use of unassigned local variable" error. `var value = a.C ?? default;` works. It's more idiomatic C# than `var value = a.C.GetValueOrDefault();` which is what it compiles to.

**Reply by kyoyama-kazusa** [reply]
Thanks.
